src/rdf4cpp/rdf/Literal.cpp

```diff
@@ -1260,6 +1408,33 @@ Literal Literal::substr(Literal const &start, Literal const &len, Node::NodeStor
     return this->substr(start_ix, len_ix, node_storage);
 }
 
+bool lang_matches(std::string_view const lang_tag, std::string_view const lang_range) noexcept {
+    if (lang_range.empty()) {
+        return lang_tag.empty();
+    }
+
+    if (lang_range == "*") {
+        return !lang_tag.empty();
+    }
+
+    auto const lang_ci = util::CiStringView{lang_tag.data(), lang_tag.size()};
+    auto const lang_range_ci = util::CiStringView{lang_range.data(), lang_range.size()};
+
+    return lang_ci.starts_with(lang_range_ci) && (lang_ci.size() == lang_range_ci.size() || lang_ci[lang_range_ci.size()] == '-');
```

Comment:
possible buffer over-read
```
lang_ci[lang_range_ci.size()] == '-'
```

Reply:
No, at least not if I am missing something (which is definitely possible), because of short-circuiting we know the following during evaluation of `lang_ci[lang_range_ci.size()] == '-'`:

1. `lang_ci.starts_with(lang_range_ci)` otherwise the second part of the `&&` would not evaluate
    - therefore `lang_range_ci.size() <= lang_ci.size()`
2. `lang_ci.size() != lang_range_ci.size()` because otherwise the second part of the `||` would not evaluate
    - therefore `lang_range_ci.size() < lang_ci.size()` (strictly less than)

therefore `lang_range_ci[lang_range_ci.size()]` is strictly in bounds


https://godbolt.org/z/szrhEh6v7

Reply:
You are right. clever :)